src/Hipace.cpp

```diff
@@ -602,60 +602,46 @@ Hipace::ExplicitSolveOneSubSlice (const int lev, const int ibox, const amrex::Bo
+    // jx_beam and jy_beam are used from the previous "Next" slice
+    // jx and jy are initially set to jx_beam and jy_beam
+    m_fields.setVal(0., lev, WhichSlice::This, "chi", "Sy", "Sx", "ExmBy", "EypBx", "Ez",
+        "Bz", "Psi", "jz_beam", "rho_beam", "jz", "rho");
 
     if (m_do_tiling) m_multi_plasma.TileSort(bx, geom[lev]);
+
+    // deposit jx, jy, jz, rho and chi for all plasmas
     m_multi_plasma.DepositCurrent(
         m_fields, m_laser, WhichSlice::This, false, true, true, true, true, geom[lev], lev);
 
     m_fields.setVal(0., lev, WhichSlice::Next, "jx_beam", "jy_beam");
-    m_multi_beam.DepositCurrentSlice(m_fields, geom, lev, islice_local, beam_bin,
-                                     m_box_sorters, ibox, m_do_beam_jx_jy_deposition,
-                                     WhichSlice::Next);
+    // deposit jx_beam and jy_beam in the Next slice
+    m_multi_beam.DepositCurrentSlice(m_fields, geom, lev, islice_local, beam_bin, m_box_sorters,
+        ibox, m_do_beam_jx_jy_deposition, false, false, WhichSlice::Next);
     // need to exchange jx_beam jy_beam
-    m_fields.FillBoundary(Geom(lev).periodicity(), lev, WhichSlice::Next,
-        "jx_beam", "jy_beam");
+    m_fields.FillBoundary(Geom(lev).periodicity(), lev, WhichSlice::Next, "jx_beam", "jy_beam");
 
     m_fields.AddRhoIons(lev);
 
-    FillBoundaryChargeCurrents(lev);
-
-    if (!m_do_beam_jz_minus_rho) {
-        m_fields.SolvePoissonExmByAndEypBx(Geom(), m_comm_xy, lev, islice);
-    }
-
-    m_multi_beam.DepositCurrentSlice(m_fields, geom, lev, islice_local, beam_bin,
-                                     m_box_sorters, ibox, m_do_beam_jx_jy_deposition,
-                                     WhichSlice::This, m_do_beam_jz_minus_rho);
-
-    if (m_do_beam_jz_minus_rho) {
-        m_fields.SolvePoissonExmByAndEypBx(Geom(), m_comm_xy, lev, islice);
-    }
-
-    m_grid_current.DepositCurrentSlice(m_fields, geom[lev], lev, islice);
+    // deposit jz_beam and maybe rho_beam on This slice
+    m_multi_beam.DepositCurrentSlice(m_fields, geom, lev, islice_local, beam_bin, m_box_sorters,
+        ibox, false, true, m_do_beam_jz_minus_rho, WhichSlice::This);
 
     FillBoundaryChargeCurrents(lev);
 
+    m_fields.SolvePoissonExmByAndEypBx(Geom(), m_comm_xy, lev, islice);
     m_fields.SolvePoissonEz(Geom(), lev, islice);
     m_fields.SolvePoissonBz(Geom(), lev, islice);
 
-    // Modifies Bx and By in the current slice and the force terms of the plasma particles
-    m_fields.AddRhoIons(lev, true);
+    // deposit grid current into jz_beam
+    m_grid_current.DepositCurrentSlice(m_fields, geom[lev], lev, islice);
+    // No FillBoundary because grid current only deposits in the middle of the field
+
+    // Modifies Bx, By, Sx and Sy in the current slice
     ExplicitSolveBxBy(lev);
-    m_multi_plasma.AdvanceParticles(m_fields, m_laser, geom[lev], false, true, true,
-                                    true, lev);
-    m_fields.AddRhoIons(lev);
```

Comment:
Previously the neutralizing background (for rho) was needed to calculate Psi, then needed to be removed for the old way of computing Sx and Sy, and lastly was added back for diagnostics. Since the field rho is now no longer needed for Sx and Sy, the second and third call to AddRhoIons can be left out.